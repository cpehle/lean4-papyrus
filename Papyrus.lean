import Papyrus.IR.Types
import Papyrus.IR.Constants
import Papyrus.IR.Module

open Papyrus

--------------------------------------------------------------------------------
-- General Test Helpers
--------------------------------------------------------------------------------

def assertEq [Repr α] [DecidableEq α] (expected : α) (actual : α) : IO PUnit := do
  unless expected = actual do
    IO.eprintln s!"expected '{repr expected}', got '{repr actual}'"

def assertBEq [Repr α] [BEq α] (expected : α) (actual : α) : IO PUnit := do
  unless expected == actual do
    IO.eprintln s!"expected '{repr expected}', got '{repr actual}'"

def testcase (name : String) [Monad m] [MonadLiftT IO m] (action : m PUnit) : m PUnit := do
  IO.println s!"Running test '{name}' ..."
  action

--------------------------------------------------------------------------------
-- Basic Module Test
--------------------------------------------------------------------------------

def testModule : LLVM PUnit := do

  testcase "omodule naming" do
    let mod ← ModuleRef.new "hello"
    assertBEq "hello" (← mod.getModuleID)
    mod.setModuleID "world"
    assertBEq "world" (← mod.getModuleID)

--------------------------------------------------------------------------------
-- Type Tests
--------------------------------------------------------------------------------

def printRefTypeID (ref : TypeRef) : IO PUnit := do
  IO.println <| repr (← ref.getTypeID)

def assertRefTypeID (expectedID : TypeID) (ref : TypeRef) : IO PUnit := do
  assertBEq expectedID (← ref.getTypeID)

def assertIntTypeRoundtrips (type : IntegerType n) : LLVM PUnit := do
  let ref ← type.getRef
  assertBEq TypeID.Integer (← ref.getTypeID)
  assertBEq n (← ref.getBitWidth).toNat

def assertFunTypeRoundtrips
[ToTypeRef r] [ToTypeRefArray p] (type : FunctionType r p a)
: LLVM PUnit := do
  let ref ← type.getRef
  assertBEq TypeID.Function (← ref.getTypeID)
  assertBEq (← (← toTypeRef type.resultType).getTypeID) (← (← ref.getReturnType).getTypeID)
  assertBEq (← toTypeRefArray type.parameterTypes).size (← ref.getParameterTypes).size
  assertBEq type.isVarArg (← ref.isVarArg)

def assertVectorTypeRoundtrips [ToTypeRef e] (type : VectorType e n s) : LLVM PUnit := do
  let ref ← type.getRef
  assertBEq (ite type.isScalable TypeID.ScalableVector TypeID.FixedVector) (← ref.getTypeID)
  assertBEq (← (← toTypeRef type.elementType).getTypeID) (← (← ref.getElementType).getTypeID)
  assertBEq type.minSize (← ref.getMinSize)
  assertBEq type.isScalable (← ref.isScalable)

def testTypes : LLVM PUnit := do

  testcase "special types" do
    assertRefTypeID TypeID.Void       (← voidType.getRef)
    assertRefTypeID TypeID.Label      (← labelType.getRef)
    assertRefTypeID TypeID.Metadata   (← metadataType.getRef)
    assertRefTypeID TypeID.Token      (← tokenType.getRef)
    assertRefTypeID TypeID.X86_MMX    (← x86MMXType.getRef)

  testcase "floating point types" do
    assertRefTypeID TypeID.Half       (← halfType.getRef)
    assertRefTypeID TypeID.BFloat     (← bfloatType.getRef)
    assertRefTypeID TypeID.Float      (← floatType.getRef)
    assertRefTypeID TypeID.Double     (← doubleType.getRef)
    assertRefTypeID TypeID.X86_FP80   (← x86FP80Type.getRef)
    assertRefTypeID TypeID.FP128      (← fp128Type.getRef)
    assertRefTypeID TypeID.PPC_FP128  (← ppcFP128Type.getRef)

  testcase "integer types" do
    assertIntTypeRoundtrips int1Type
    assertIntTypeRoundtrips int8Type
    assertIntTypeRoundtrips int16Type
    assertIntTypeRoundtrips int32Type
    assertIntTypeRoundtrips int64Type
    assertIntTypeRoundtrips int128Type
    assertIntTypeRoundtrips <| integerType 100

  testcase "function types" do
    assertFunTypeRoundtrips <| functionType voidType doubleType
    assertFunTypeRoundtrips <| functionType voidType (floatType, int1Type) true

  testcase "pointer types" do
    let ref ← doubleType.pointerType.getRef
    assertRefTypeID TypeID.Pointer ref
    assertRefTypeID TypeID.Double (← ref.getPointeeType)
    assertBEq AddressSpace.default (← ref.getAddressSpace)

  testcase "literal struct types" do
    let ref ← literalStructType (halfType, doubleType) |>.getRef
    assertRefTypeID TypeID.Struct ref
    assertBEq 2 (← ref.getElementsTypes).size
    assertBEq false (← ref.isPacked)

  testcase "complete struct types" do
    let name := "foo"
    let ref ← completeStructType name halfType true |>.getRef
    assertRefTypeID TypeID.Struct ref
    assertBEq name (← ref.getName)
    assertBEq 1 (← ref.getElementsTypes).size
    assertBEq true (← ref.isPacked)

  testcase "opaque struct types" do
    let name := "bar"
    let ref ← opaqueStructType name |>.getRef
    assertRefTypeID TypeID.Struct ref
    assertBEq name (← ref.getName)

  testcase "array types" do
    let size := 8
    let ref ← arrayType halfType size |>.getRef
    assertRefTypeID TypeID.Array ref
    assertRefTypeID TypeID.Half (← ref.getElementType)
    assertBEq size (← ref.getSize)

  testcase "vector types" do
    assertVectorTypeRoundtrips <| fixedVectorType doubleType 8
    assertVectorTypeRoundtrips <| scalableVectorType floatType 8

--------------------------------------------------------------------------------
-- Constant Tests
--------------------------------------------------------------------------------

def testConstants : LLVM PUnit := do

  let int8TypeRef ← int8Type.getRef
  let int128TypeRef ← int128Type.getRef

  testcase "big null integer constant" do
    let const : ConstantIntRef ← int128TypeRef.getNullConstant
    assertBEq 0 (← const.getNatValue)
    assertBEq 0 (← const.getValue)

  testcase "big all ones integer constant" do
    let const : ConstantIntRef ← int128TypeRef.getAllOnesConstant
    assertBEq (2 ^ 128 - 1) (← const.getNatValue)
    assertBEq (-1) (← const.getValue)

  testcase "small positive constructed integer constant" do
    let val := 32
    let const ← int8TypeRef.getConstantInt val
    assertBEq val (← const.getNatValue)
    assertBEq val (← const.getValue)

  testcase "small negative constructed integer constant" do
    let absVal := 32; let intVal := -32
    let const ← int8TypeRef.getConstantInt intVal
    assertBEq (2 ^ 8 - absVal) (← const.getNatValue)
    assertBEq intVal (← const.getValue)

  testcase "big positive constructed integer constant" do
    let val : Nat := 2 ^ 80 + 12
    let const ← int128TypeRef.getConstantInt val
    assertBEq (Int.ofNat val) (← const.getValue)
    assertBEq val (← const.getNatValue)

  testcase "big negative constructed integer constant" do
    let absVal := 2 ^ 80 + 12
    let intVal := -(Int.ofNat absVal)
    let const ← int128TypeRef.getConstantInt intVal
    assertBEq (Int.ofNat (2 ^ 128) - absVal) (← const.getNatValue)
    assertBEq intVal (← const.getValue)

--------------------------------------------------------------------------------
-- Test Runner
--------------------------------------------------------------------------------

def main : IO PUnit := LLVM.run do

  testTypes
  testConstants
  testModule

  IO.println "All tests finished."
