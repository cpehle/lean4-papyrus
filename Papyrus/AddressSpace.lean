namespace Papyrus

/-- A numerically indexed address space. -/
structure AddressSpace where
  index : Nat

namespace AddressSpace

/-- The default address space (i.e., 0). -/
def default := mk 0

/-- Make an address space from a `Nat`. -/
def ofNat (n : Nat) := mk n

end AddressSpace

instance : Inhabited AddressSpace := ⟨AddressSpace.default⟩
instance {n} : OfNat AddressSpace n := ⟨AddressSpace.ofNat n⟩
