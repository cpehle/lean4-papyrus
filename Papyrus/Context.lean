namespace Papyrus

constant ContextRef : Type := Unit

@[extern "papyrus_context_new"]
constant ContextRef.new : IO ContextRef

/-- The LLVM Monad. -/
abbrev LLVM := ReaderT ContextRef IO

namespace LLVM

def runWith (ctx : ContextRef) (self : LLVM α) : IO α :=
  self ctx

def run (self : LLVM α) : IO α := do
  self (← ContextRef.new)
