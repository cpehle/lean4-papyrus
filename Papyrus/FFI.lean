namespace Papyrus

/-- An object which defines the kind of an external pointer. -/
constant ExternalPtrClass : Type := Unit

/-- A Lean object which contains an external pointer. -/
constant ExternalPtr : ExternalPtrClass → Type → Type := fun _ _ => Unit

/- The external class of LoosePtr. -/
axiom LoosePtr.class : ExternalPtrClass

/-- An external pointer with no memory management. -/
abbrev LoosePtr (α) := ExternalPtr Papyrus.LoosePtr.class α

/- The external class of LoosePtr. -/
axiom OwnedPtr.class : ExternalPtrClass

/--
  A external pointer that is deleted upon being garbage collected
  (i.e., it is owned by Lean).
-/
abbrev OwnedPtr (α) := ExternalPtr Papyrus.OwnedPtr.class α

/-- The actual implementation of `LinkedPtr`. -/
structure LinkedPtrImpl (k : ExternalPtrClass) (σ) (α) where
  link : σ
  this : ExternalPtr k α

/--
  A external pointer whose lifetime is linked to some other external object
  that should not be deleted until this one is garbage collected by Lean.

  It as an opaque definition to prevent access to the internal `ExternalPtr`,
  which could create memory mismanagement (e.g., if a reference to it is kept
  without keeping a reference to the `LinkedPtr`).
-/
constant LinkedPtr (k : ExternalPtrClass) (σ : Type) (α : Type) : Type :=
  LinkedPtrImpl k σ α

/--  A linked external pointer that should not be managed by Lean. -/
abbrev LinkedLoosePtr := LinkedPtr LoosePtr.class

/--  A linked external pointer that can be deleted independently by Lean. -/
abbrev LinkedOwnedPtr := LinkedPtr OwnedPtr.class
