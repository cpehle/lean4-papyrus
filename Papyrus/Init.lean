namespace Papyrus

-- # All Target Initialization

/-- Initializes all LLVM supported targets. -/
@[extern "papyrus_init_all_targets"]
constant initAllTargets : IO PUnit

/-- Initializes all LLVM supported target MCs. -/
@[extern "papyrus_init_all_target_mcs"]
constant initAllTargetMCs : IO PUnit

/-- Initializes all LLVM supported target infos. -/
@[extern "papyrus_init_all_target_infos"]
constant initAllTargetInfos : IO PUnit

/-- Initializes all LLVM supported target assembly parsers. -/
@[extern "papyrus_init_all_asm_parsers"]
constant initAllAsmParsers : IO PUnit

/-- Initializes all LLVM supported target assembly printers. -/
@[extern "papyrus_init_all_asm_printers"]
constant initAllAsmPrinters : IO PUnit

/-- Initializes all LLVM supported target disassemblers. -/
@[extern "papyrus_init_all_disassemblers"]
constant initAllDisassemblers : IO PUnit

-- # Native Target Initialization

/--
  Initializes the native target along with its MC and Info.
  Returns true if no native target exists, false otherwise.
-/
@[extern "papyrus_init_native_target"]
constant initNativeTarget : IO Bool

/--
  Initializes the native target's ASM parser.
  Returns true if no native target exists, false otherwise.
-/
@[extern "papyrus_init_native_asm_parser"]
constant initNativeAsmParser : IO Bool


/--
  Initializes the native target's ASM printer.
  Returns true if no native target exists, false otherwise.
-/
@[extern "papyrus_init_native_asm_printer"]
constant initNativeAsmPrinter : IO Bool


/--
  Initializes the native target's disassembler.
  Returns true if no native target exists, false otherwise.
-/
@[extern "papyrus_init_native_disassembler"]
constant initNativeDisassembler : IO Bool
