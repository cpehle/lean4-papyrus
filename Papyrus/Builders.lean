import Papyrus.Context
import Papyrus.IR.ModuleRef
import Papyrus.IR.FunctionRef
import Papyrus.IR.GlobalVariableRef
import Papyrus.IR.InstructionRefs

namespace Papyrus

-- # Module Builder

abbrev ModuleM := ReaderT ModuleRef LlvmM

protected def ModuleM.runIn (mod : ModuleRef) (self : ModuleM α) : LlvmM α :=
  self mod

-- # Basic Block Builder

structure BasicBlockContext where
  modRef : ModuleRef
  funRef : FunctionRef
  bbRef : BasicBlockRef

abbrev BasicBlockM := ReaderT BasicBlockContext LlvmM

instance : MonadLift ModuleM BasicBlockM where
  monadLift m := fun ctx => m ctx.modRef

protected def BasicBlockM.runIn (ctx : BasicBlockContext) (self : BasicBlockM α) : LlvmM α :=
  self ctx

namespace Actions

def module (name : String) (builder : ModuleM PUnit) : LlvmM ModuleRef := do
  let modRef ← ModuleRef.new name
  builder.runIn modRef
  return modRef

-- ## Module Actions

/-- Add a global variable to the module. -/
def globalVar (type : TypeRef)
(isConstant := false) (linkage := Linkage.external) (name : @& String := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default) (isExternallyInitialized := false)
: ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.new type isConstant linkage name tlm addrSpace isExternallyInitialized
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a global variable with an initializer to the module. -/
def globalVarInit (type : TypeRef)
(isConstant := false) (linkage := Linkage.external) (name : @& String := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default)
: ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.new type isConstant linkage name tlm addrSpace false
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a string constant to the module. -/
def string (str : String)
(name := "") (addrSpace := AddressSpace.default) (withNull := true) : ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.ofString str name addrSpace withNull
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a string constant to the module and return a constant pointer to its head. -/
def stringPtr (str : String)
(name := "") (addrSpace := AddressSpace.default) (withNull := true) : ModuleM ConstantRef := do
  let gblRef ← string str name addrSpace withNull
  let zeroRef ← ConstantWordRef.ofUInt32 0
  let ptrRef ← ConstantExprRef.getGetElementPtr gblRef #[zeroRef, zeroRef] true
  return ptrRef

/-- Add a arbitrary constant to the module. -/
def globalConst (init : ConstantRef)
(linkage := Linkage.linkOnceODR) (name := "")
(tlm := ThreadLocalMode.notLocal) (addrSpace := AddressSpace.default) : ModuleM GlobalVariableRef := do
  let gblRef ← GlobalVariableRef.ofConstant init true linkage name tlm addrSpace
  (← read).appendGlobalVariable gblRef
  return gblRef

/-- Add a function declaration to the module. -/
def declare (type : FunctionTypeRef) (name : String)
(linkage := Linkage.external) (addrSpace := AddressSpace.default) : ModuleM FunctionRef := do
  let funRef ← FunctionRef.create type name linkage addrSpace
  (← read).appendFunction funRef
  return funRef

/-- Define a new a function in the module. -/
def define (type : FunctionTypeRef) (builder : BasicBlockM PUnit) (name : String :=  "")
(linkage := Linkage.external) (addrSpace := AddressSpace.default) (entry : String := "") : ModuleM FunctionRef := do
  let funRef ← FunctionRef.create type name linkage addrSpace
  let bbRef ← BasicBlockRef.create entry
  funRef.appendBasicBlock bbRef
  let modRef ← read
  modRef.appendFunction funRef
  builder.runIn {modRef, funRef, bbRef}
  return funRef

-- ## Basic Block Actions

def call (fn : FunctionRef) (args : Array ValueRef := #[]) (name : String := "") : BasicBlockM InstructionRef := do
  let inst ← fn.createCall args name
  (← read).bbRef.appendInstruction inst
  return inst

def callAs (type : FunctionTypeRef) (fn : ValueRef) (args : Array ValueRef := #[]) (name : String := "") : BasicBlockM InstructionRef := do
  let inst ← CallInstRef.create type fn args name
  (← read).bbRef.appendInstruction inst
  return inst

def retVoid : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← ReturnInstRef.createVoid

def ret (val : ValueRef)  : BasicBlockM PUnit := do
  (← read).bbRef.appendInstruction <| ← ReturnInstRef.create val
