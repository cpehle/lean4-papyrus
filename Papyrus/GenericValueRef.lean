import Papyrus.FFI
import Papyrus.IR.TypeRefs

namespace Papyrus

/--
  An opaque type representing an external
  [GenericValue](https://llvm.org/doxygen/structllvm_1_1GenericValue.html).
-/
constant Llvm.GenericValue : Type := Unit

/--
  A reference to an external LLVM
  [GenericValue](https://llvm.org/doxygen/structllvm_1_1GenericValue.html).
-/
def GenericValueRef := OwnedPtr Llvm.GenericValue

namespace  GenericValueRef

/--
  Create a integer generic of the given width with the given `Int` value.
  The value will be truncated and/or extended as necessary to make it fit.
-/
@[extern "papyrus_generic_value_of_int"]
constant ofInt (numBits : UInt32) (value : @& Int) : IO GenericValueRef

/-- Get the value of this generic as an `Int` by treating its integer bits as signed. -/
@[extern "papyrus_generic_value_to_int"]
constant toInt (self : @& GenericValueRef) : IO Int

/--
  Create a integer generic  of the given width with the given `Nat` value.
  The value will be truncated and/or extended as necessary to make it fit.
-/
@[extern "papyrus_generic_value_of_nat"]
constant ofNat (numBits : UInt32) (value : @& Nat)  : IO GenericValueRef

/-- Get the integer value of this generic as a `Nat` by treating its integer bits as unsigned. -/
@[extern "papyrus_generic_value_to_nat"]
constant toNat (self : @& GenericValueRef) : IO Nat

/-- Create a `double` generic from a `Float`. -/
@[extern "papyrus_generic_value_of_float"]
constant ofFloat (value : @& Float) : IO GenericValueRef

/-- Get the `double` value of this generic as a `Float`. -/
@[extern "papyrus_generic_value_to_float"]
constant toFloat (self : @& GenericValueRef) : IO Float

/-- Create an aggregate generic from an `Array`. -/
@[extern "papyrus_generic_value_of_array"]
constant ofArray (value : @& Array GenericValueRef) : IO GenericValueRef

/-- Get the aggregate value of this generic as an `Array`. -/
@[extern "papyrus_generic_value_to_array"]
constant toArray (self : @& GenericValueRef) : IO (Array GenericValueRef)

end GenericValueRef

namespace IntegerTypeRef

/-- Get a reference to a generic of this type with the value of `Int`. -/
def getGenericValueOfInt (value : @& Int) (self : @& IntegerTypeRef) : IO GenericValueRef := do
  GenericValueRef.ofInt (← self.getBitWidth) value

/-- Get a reference to a generic of this type with the value of `Nat`. -/
constant getGenericValueOfNat (value : @& Nat) (self : @& IntegerTypeRef) : IO GenericValueRef := do
  GenericValueRef.ofNat (← self.getBitWidth) value

end IntegerTypeRef
