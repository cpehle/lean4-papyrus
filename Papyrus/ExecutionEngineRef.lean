import Papyrus.IR.ModuleRef
import Papyrus.IR.FunctionRef
import Papyrus.GenericValueRef

namespace Papyrus

/-- A desired execution engine kind. -/
inductive EngineKind
| Either
| JIT
| Interpreter
deriving BEq, Repr

attribute [unbox] EngineKind
instance : Inhabited EngineKind := ⟨EngineKind.Either⟩

namespace EngineKind

def isJIT (self : EngineKind) : Bool :=
  self matches JIT

def allowsJIT : (self : EngineKind) → Bool
| JIT => true
| Interpreter => false
| Either => true

def isInterpreter (self : EngineKind) : Bool :=
  self matches Interpreter

def allowsInterpreter : (self : EngineKind) → Bool
| JIT => false
| Interpreter => true
| Either => true

def isEither (self : EngineKind) : Bool :=
  self matches Either

def ofUInt8 (value : UInt8) : EngineKind :=
  if value == 0b01 then JIT else
  if value == 0b10 then Interpreter else
  Either

def toUInt8 : (self : EngineKind) → UInt8
| JIT => 0b01
| Interpreter => 0b10
| Either => 0b11

end EngineKind

/-- Optimization level for generated code. -/
inductive OptLevel
| /-- -O0 -/ None
| /-- -O1 -/ Less
| /-- -O2 -/ Default
| /-- -O3 -/ Aggressive
deriving BEq, Repr

attribute [unbox] OptLevel
instance : Inhabited OptLevel := ⟨OptLevel.Default⟩

/--
  A reference to an external LLVM
  [ExecutionEngine](https://llvm.org/doxygen/classllvm_1_1ExecutionEngine.html).
-/
constant ExecutionEngineRef : Type := Unit

namespace ExecutionEngineRef

/-- Create an execution engine for the given module. -/
@[extern "papyrus_execution_engine_create_for_module"]
constant createForModule (mod : @& ModuleRef) (kind : @& EngineKind := EngineKind.Either)
  (march : @& String := "") (mcpu : @& String := "") (mattrs : @& Array String := #[])
  (optLevel : @& OptLevel := OptLevel.Default) (verifyModule := false)
  : IO ExecutionEngineRef

/--
  Execute the given function with the given arguments, and return the result.

  MCJIT execution engines can only execute 'main-like' function.
  That is, those returning `void` or `int` and taking no arguments
  (i.e., `[]`) or argc/argv (i.e., `[int, char**]`).
-/
@[extern "papyrus_execution_engine_run_function"]
constant runFunction (fn : @& FunctionRef) (args : @& Array GenericValueRef)
  (self : @& ExecutionEngineRef) : IO GenericValueRef

end ExecutionEngineRef
