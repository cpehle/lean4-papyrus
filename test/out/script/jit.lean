import Papyrus

open Papyrus Script

llvm module exit do
  define i32 @main() do
   ret i32 101

#jit exit

llvm module echo do
  define i32 @main(argc : i32) do
   ret %argc

#jit echo #["a", "b", "c"]

llvm module empty do
  pure ()

#jit empty -- Error: Module has no main function

llvm module lean_hello do
  declare %lean_object* @lean_mk_string(i8*)
  declare %lean_object* @l_IO_println___at_Lean_instEval___spec__1(%lean_object*, %lean_object*);
  define i32 @main() do
   %hello = call @lean_mk_string("Hello World!"*)
   call @l_IO_println___at_Lean_instEval___spec__1(%hello, %hello)
   ret i32 0

#jit lean_hello
