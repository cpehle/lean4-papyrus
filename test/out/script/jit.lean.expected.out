101
3

out/script/jit.lean:20:0-20:10: error: Module has no main function
Hello World!
0
